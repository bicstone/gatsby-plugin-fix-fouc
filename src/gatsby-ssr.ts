import * as React from "react";

import { camelCase } from "camel-case";
import { defaultOptions } from ".";
import type { GatsbyPluginFixFoucOptions } from ".";

import type { RenderBodyArgs } from "gatsby";

export const onRenderBody = (
  { setHeadComponents, setBodyAttributes }: RenderBodyArgs,
  pluginOptions: GatsbyPluginFixFoucOptions
): void => {
  const attributeName =
    pluginOptions.attributeName ?? defaultOptions.attributeName;
  const minWidth = pluginOptions.minWidth ?? defaultOptions.minWidth;
  const timeout = pluginOptions.timeout ?? defaultOptions.timeout;

  setHeadComponents([
    React.createElement("style", {
      key: "loading-screen-style",
      dangerouslySetInnerHTML: {
        __html:
          `body[data-${attributeName}]{opacity:1}` +
          `@media(min-width:${minWidth}px){body[data-${attributeName}]{opacity:0}}`,
      },
    }),

    React.createElement("noscript", {
      key: "loading-screen-noscript-style",
      dangerouslySetInnerHTML: {
        __html:
          `<style>` +
          `body[data-${attributeName}]{opacity:1!important}` +
          `</style>`,
      },
    }),

    React.createElement("script", {
      key: "loading-screen-fail-safe",
      dangerouslySetInnerHTML: {
        __html:
          `(function(){` +
          `setTimeout(function(){` +
          `try{` +
          `if(` +
          `document.body.dataset["${camelCase(attributeName)}"]!==undefined` +
          `){` +
          `delete document.body.dataset["${camelCase(attributeName)}"]` +
          `}` +
          `}catch(e){}` +
          `},${timeout})` +
          `})();`,
      },
    }),
  ]);

  setBodyAttributes({
    [`data-${attributeName}`]: "true",
  });
};
