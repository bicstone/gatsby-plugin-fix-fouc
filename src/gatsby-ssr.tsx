import React from "react";

import camelCase from "camelcase";
import { defaultOptions, GatsbyPluginFixFoucOptions } from "./";

import type { RenderBodyArgs } from "gatsby";

const generateHtml = (
  str: string
): React.DOMAttributes<Element>["dangerouslySetInnerHTML"] => {
  return {
    __html: str.trim().replace("/\n/g", ""),
  };
};

export const onRenderBody = (
  { setHeadComponents, setBodyAttributes }: RenderBodyArgs,
  pluginOptions: GatsbyPluginFixFoucOptions
): void => {
  const attributeName =
    pluginOptions.attributeName ?? defaultOptions.attributeName;
  const minWidth = pluginOptions.minWidth ?? defaultOptions.minWidth;
  const timeout = pluginOptions.timeout ?? defaultOptions.timeout;

  setHeadComponents([
    <style
      key="loading-screen-style"
      dangerouslySetInnerHTML={generateHtml(`
        body[data-${attributeName}]{opacity:1}
        @media(min-width:${minWidth}px){body[data-${attributeName}]{opacity:0}}
      `)}
    />,

    <noscript
      key="loading-screen-noscript-style"
      dangerouslySetInnerHTML={generateHtml(`
      <style>
        body[data-${attributeName}]{opacity:1!important}
      </style>
      `)}
    />,
    <script
      key="loading-screen-fail-safe"
      dangerouslySetInnerHTML={generateHtml(`
        setTimeout(function(){
          delete document.body.dataset?.["${camelCase(attributeName)}"]
        },${timeout})
      `)}
    />,
  ]);

  setBodyAttributes({
    [`data-${attributeName}`]: "true",
  });
};
