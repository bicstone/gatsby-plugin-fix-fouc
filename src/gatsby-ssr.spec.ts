/**
 * @jest-environment jsdom
 */

import { defaultOptions } from ".";
import { onRenderBody } from "./gatsby-ssr";

const renderBodyArgs = {
  loadPageDataSync: jest.fn(),
  pathname: "",
  setHeadComponents: jest.fn(),
  setHtmlAttributes: jest.fn(),
  setBodyAttributes: jest.fn(),
  setPreBodyComponents: jest.fn(),
  setPostBodyComponents: jest.fn(),
  setBodyProps: jest.fn(),
};

describe(`onRenderBody`, () => {
  describe(`TODO`, () => {
    test(`set head components`, () => {
      // const attributeName = defaultOptions.attributeName;

      onRenderBody(renderBodyArgs, {});

      expect(renderBodyArgs.setHeadComponents).toHaveBeenCalledTimes(1);
      expect(renderBodyArgs.setHeadComponents).toMatchSnapshot();
    });

    test(`set head components (use attributeName option)`, () => {
      const attributeName = "is-loading";

      onRenderBody(renderBodyArgs, { attributeName });

      expect(renderBodyArgs.setBodyAttributes).toHaveBeenCalledTimes(1);
      expect(renderBodyArgs.setBodyAttributes).toMatchSnapshot();
    });
  });

  describe(`setBodyAttributes`, () => {
    test(`sets body attributes`, () => {
      const attributeName = defaultOptions.attributeName;

      onRenderBody(renderBodyArgs, {});

      expect(renderBodyArgs.setBodyAttributes).toHaveBeenCalledTimes(1);
      expect(renderBodyArgs.setBodyAttributes).toHaveBeenCalledWith({
        [`data-${attributeName}`]: "true",
      });
    });

    test(`sets body attributes (use attributeName option)`, () => {
      const attributeName = "is-loading";

      onRenderBody(renderBodyArgs, { attributeName });

      expect(renderBodyArgs.setBodyAttributes).toHaveBeenCalledTimes(1);
      expect(renderBodyArgs.setBodyAttributes).toHaveBeenCalledWith({
        [`data-${attributeName}`]: "true",
      });
    });
  });
});
