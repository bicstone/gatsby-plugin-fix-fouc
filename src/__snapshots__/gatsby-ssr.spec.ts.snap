// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onRenderBody TODO set head components (use attributeName option) 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "data-is-loading": "true",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`onRenderBody TODO set head components 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        <style
          dangerouslySetInnerHTML={
            {
              "__html": "body[data-gatsby-plugin-fix-fouc-is-loading]{opacity:1}
        @media(min-width:0px){body[data-gatsby-plugin-fix-fouc-is-loading]{opacity:0}}",
            }
          }
        />,
        <noscript
          dangerouslySetInnerHTML={
            {
              "__html": "<style>
        body[data-gatsby-plugin-fix-fouc-is-loading]{opacity:1!important}
      </style>",
            }
          }
        />,
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "setTimeout(function(){
          delete document.body.dataset?.["gatsbyPluginFixFoucIsLoading"]
        },2000)",
            }
          }
        />,
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
